components/pluralrules/src/rules/ast.rs

```diff
@@ -0,0 +1,436 @@
+//! to evaluate whether a given [`PluralCategory`] should be used.
+//!
+//! # Examples
+//!
+//! ```
+//! use icu_pluralrules::rules::parse;
+//! use icu_pluralrules::rules::ast::*;
+//!
+//! let input = "i = 1";
+//!
+//! let ast = parse(input.as_bytes())
+//!     .expect("Parsing failed.");
+//!
+//! assert_eq!(ast, Condition(Box::new([
+//!     AndCondition(Box::new([
+//!         Relation {
+//!             expression: Expression {
+//!                 operand: Operand::I,
+//!                 modulus: None,
+//!             },
+//!             operator: Operator::Eq,
+//!             range_list: RangeList(Box::new([
+//!                 RangeListItem::Value(
+//!                     Value(1)
+//!                 )
+//!             ]))
+//!         }
+//!     ]))
+//! ])));
+//! ```
+//!
+//! [`PluralCategory`]: ../enum.PluralCategory.html
+//! [`parse`]: ../fn.parse.html
+//! [`test_condition`]: ../fn.test_condition.html
+use std::ops::RangeInclusive;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Rule {
+    pub condition: Condition,
+    pub samples: Option<Samples>,
+}
+
+/// A complete (and the only complete) AST representation of a plural rule. Comprises a vector of AndConditions.
+///
+/// # Examples
+///
+/// All AST nodes can be built explicitly, as seen in the example. However, due to its complexity, it is preferred to build the AST using the parse_plural_rule function.
+///
+/// ```text
+/// "i = 5"
+/// ```
+///
+/// Can be represented by the AST:
+///
+/// ```
+/// use icu_pluralrules::rules::ast::*;
+///
+/// Condition(Box::new([AndCondition(Box::new([Relation {
+///        expression: Expression {
+///            operand: Operand::I,
+///            modulus: None,
+///        },
+///        operator: Operator::Eq,
+///        range_list: RangeList(Box::new([RangeListItem::Value(Value(5))])),
+///    }]))]));
+/// ```
+///
+/// Because they care complete representations, hand-written Conditions can be verified with the assert macro. No other AST nodes can be verified.
+///
+/// ```
+/// use icu_pluralrules::rules::ast::*;
+/// use icu_pluralrules::rules::parse;
+///
+/// let condition = Condition(Box::new([
+///     AndCondition(Box::new([Relation {
+///         expression: Expression {
+///             operand: Operand::I,
+///             modulus: None,
+///         },
+///         operator: Operator::Eq,
+///         range_list: RangeList(Box::new([RangeListItem::Value(Value(5))])),
+///     }])),
+///     AndCondition(Box::new([Relation {
+///         expression: Expression {
+///             operand: Operand::V,
+///             modulus: None,
+///         },
+///         operator: Operator::Eq,
+///         range_list: RangeList(Box::new([RangeListItem::Value(Value(2))])),
+///     }])),
+/// ]));
+///
+/// assert_eq!(
+///     condition,
+///     parse(b"i = 5 or v = 2")
+///          .expect("Parsing failed")
+/// )
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct Condition(pub Box<[AndCondition]>);
```

Comment:
You have a lot of boxes, which means a lot of mallocs.  **This is probably fine**, but I wonder if smallvec would be useful, since I expect that most of these lists will be between 0 and 2 elements long, with a few going up to 5 or 6.

Reply:
good point! I'll experiment with this next!

Reply:
I did experiment and I see inconclusive results (criterion `parse` improves, but intl-measurements harness slightly regresses). I'm going to file an issue to explore that avenue separately from this PR.

Reply:
Can you link the follow-up issue?

Reply:
https://github.com/unicode-org/icu4x/issues/193